python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_restraint.py

```diff
@@ -21,4 +21,198 @@
 
 """
 A class for holding restraints.
-"""
\ No newline at end of file
+"""
+
+from .._SireWrappers import Atom
+from ..Types import Length, Angle
+from .._SireWrappers import System as _System
+from ..Units.Length import nanometer
+from ..Units.Angle import degree, radian
+from ..Units.Energy import kj_per_mol
+
+class Restraint():
+    '''The Restraint class which holds the restraint information for the ABFE
+    calculations. Currently only Boresch type restraint is supported.
+
+    Boresch restraint is a set of harmonic restraints containing one bond, two
+    angle and three dihedrals, which comes from three atoms in the ligand
+    (l1, l2, l3) and three atoms in the protein (r1, r2, r3). The restraints
+    are arranged in the format of atom1-atom2 (equilibrium value, force constant):
+
+    Bonds: r1-l1 (r0, kr)
+    Angles: r2-r1-l1 (thetaA0, kthetaA), r1-l1-l2 (thetaB0, kthetaB)
+    Dihedrals: r3-r2-r1-l1 (phiA0, kphiA), r2-r1-l1-l2 (phiB0, kphiB), r1-l1-l2-l3 (phiC0, kphiC)
+
+    '''
+    def __init__(self, system, restraint_dict, type='Boresch'):
```

Comment:
It might be better to avoid passing in the system with the restraint, if we are planning to do this:
```
restraint , best_frame = restraint_search.analyse() # where restraint is of the type BSS.FreeEnergy.Restraint
protocol = BSS.Protocol.FreeEnergy(restraint=restraint, ...) 
ABFE = BSS.FreeEnergy.Absolute(system, protocol, work_dir="output") # Handle writing restraints here based on the engine
```
Because storing the system in the restraint class would mean effectively passing in the system twice to BSS.FreeEnergy.Absolute . Maybe it would be more robust to take the system as an argument to toString, so in BSS.FreeEnergy.Absolute(system, protocol, work_dir="output") (or relative), only one system is considered, and there can never be any issues with differing systems. BSS.FreeEnergy.Absolute could then call Restraint.toString(system,...) internally.

Reply:
> Because storing the system in the restraint class would mean effectively passing in the system twice to BSS.FreeEnergy.Absolute.

This is a tricky issue, I do agree that passing the system twice is not optimal but I think passing `system` to `toString` is not intuitive.

I realised that I forget to implement this but I will also need to check if the ligand atoms are in the decoupled molecule, which is a property of the system.

Reply:
I have give it some thought but I cannot think of an easy way out.

So in the later stage, the BSS.FreeEnergy.Restraint object will be part of the BSS._SireWrappers.System. This should take care of the topology writer issue as well. However, this could be a large task and might take some time.

My vision of the current version of the BSS.FreeEnergy.Restraint object is that it will be a patch to the System to hold the Restraint.

With regard to the issue of `maybe starting from different conformations without bothering to run another RestraintSearch` 
One solution could be that give BSS.FreeEnergy.Restraint a method update_system(), to update the underlaying system without changing the restraint.